tests/unit/test_vcs.py

```diff
@@ -108,37 +109,89 @@ def test_looks_like_hash(sha, expected):
 
 
 @pytest.mark.parametrize('vcs_cls, remote_url, expected', [
-    # Git is one of the subclasses using the base class implementation.
-    (Git, 'git://example.com/MyProject', False),
+    # Mercurial is one of the subclasses using the base class implementation.
+    # `hg://` isn't a real prefix but it tests the default behaviour.
+    (Mercurial, 'hg://user@example.com/MyProject', False),
+    (Mercurial, 'http://example.com/MyProject', True),
+    # The Git subclasses should return true in all cases.
+    (Git, 'git://example.com/MyProject', True),
     (Git, 'http://example.com/MyProject', True),
-    # Subversion is the only subclass overriding the base class implementation.
+    # Subversion also overrides the base class implementation.
     (Subversion, 'svn://example.com/MyProject', True),
 ])
 def test_should_add_vcs_url_prefix(vcs_cls, remote_url, expected):
     actual = vcs_cls.should_add_vcs_url_prefix(remote_url)
     assert actual == expected
 
 
+@pytest.mark.parametrize("url, target", [
+    # A fully qualified remote url. No changes needed.
+    ("ssh://bob@server/foo/bar.git", "ssh://bob@server/foo/bar.git"),
+    ("git://bob@server/foo/bar.git", "git://bob@server/foo/bar.git"),
+    # User is optional and does not need a default.
+    ("ssh://server/foo/bar.git", "ssh://server/foo/bar.git"),
+    # The common scp shorthand for ssh remotes. Pip won't recognise these as
+    # git remotes until they have a 'ssh://' prefix and the ':' in the middle
+    # is gone.
+    ("git@example.com:foo/bar.git", "ssh://git@example.com/foo/bar.git"),
+    ("example.com:foo.git", "ssh://example.com/foo.git"),
+    # Http(s) remote names are already complete and should remain unchanged.
+    ("https://example.com/foo", "https://example.com/foo"),
+    ("http://example.com/foo/bar.git", "http://example.com/foo/bar.git"),
+    ("https://bob@example.com/foo", "https://bob@example.com/foo"),
+ ])
+def test_git_remote_url_to_pip(url, target):
+    assert Git._git_remote_to_pip_url(url) == target
+
+
+@pytest.mark.parametrize("url", [
+    # Windows paths with the ':' drive prefix look dangerously close to SCP.
+    "c:/piffle/wiffle/waffle/poffle.git",
+    r"c:\faffle\waffle\woffle\piffle.git"
```

Comment:
Will this case not pass on non-Windows?

```suggestion
    pytest.param(
        r"c:\faffle\waffle\woffle\piffle.git",
        marks=pytest.mark.skipif(
            os.name != "nt",
            reason="Windows-style path only allowed on Windows",
        ),
    ),
```

Reply:
I deliberately wanted to test that on all platforms because it's key to verifying the SCP regex. Namely that the regex pattern must not match Windows paths. It works on posix too but it's taking advantage of the fact that `os.path.exists("an invalid path")` silently returns false without complaint.

I'll change the test to check the regex directly. That way it's not making any assumptions about `os.path.exists()` and still allows a non-windows developer to properly verify the regex instantly with `pytest tests/unit/test_vcs.py` rather than having to wait 15 minutes for Windows CI to do so.

Reply:
I eventually opted for doing both. i.e. check that all paths do not match the regex then only verify that non matches propagate to exceptions using native paths.

https://github.com/pypa/pip/pull/9822/files#diff-00594ea60691593e055280834dbd177bc74f046a3bf8cb16a7822aa4bc7051b9R148-R162